

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Mouse Visium Kidney &mdash; Giotto 1.0 documentation</title>
  

  
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/copybutton.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/css/theme_edits.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/design-style.d53ab927f7bad3eaa42da95908504b10.min.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/css/theme_edits.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script src="../../_static/clipboard.min.js"></script>
        <script src="../../_static/copybutton.js"></script>
        <script src="../../_static/design-tabs.js"></script>
    
    <script type="text/javascript" src="../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Mouse osmFISH SS Cortex" href="osmFISH_mouse_SS_cortex.html" />
    <link rel="prev" title="Mouse Visium Kidney" href="mouse_visium_kidney.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../index.html">
          

          
            
            <img src="../../_static/GiottoLogo.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption" role="heading"><span class="caption-text">General</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../introduction.html">About</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../gettingstarted.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../documentation.html">Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../development.html">Development</a></li>
<li class="toctree-l1"><a class="reference external" href="http://spatial.rc.fas.harvard.edu/giotto-viewer/">Giotto Viewer</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Guides</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../WorkflowAndHowTos.html">Workflow</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../datasets.html">Examples</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../../datasets.html#dataset-information">Dataset Information</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../../datasets.html#example-datasets">Example Datasets</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="../../datasets.html#mini-datasets">Mini Datasets</a></li>
<li class="toctree-l3 current"><a class="reference internal" href="../../datasets.html#full-datasets">Full Datasets</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="seqFISH_cortex.html">    Mouse seqFISH Cortex</a></li>
<li class="toctree-l4"><a class="reference internal" href="merFISH_hypot_preopt_region.html">Mouse merFISH Hypoth. Preopt. Region</a></li>
<li class="toctree-l4"><a class="reference internal" href="STARmap_mouse_cortex.html">Mouse STARmap Cortex</a></li>
<li class="toctree-l4"><a class="reference internal" href="mouse_visium_brain.html">    Mouse Visium Brain</a></li>
<li class="toctree-l4"><a class="reference internal" href="mouse_visium_kidney.html">    Mouse Visium Kidney</a></li>
<li class="toctree-l4 current"><a class="current reference internal" href="#">    Mouse CODEX Spleen</a></li>
<li class="toctree-l4"><a class="reference internal" href="osmFISH_mouse_SS_cortex.html">    Mouse osmFISH SScortex</a></li>
<li class="toctree-l4"><a class="reference internal" href="human_CyCIF_PDAC.html">    Human CyCIF PDAC</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../trygiotto.html">Try Giotto</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Issues</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../errorsfaqsandtips.html">Common Errors and Solutions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../github_issues.html">Report a Bug</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">Giotto</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../../datasets.html">Examples</a> &raquo;</li>
        
      <li>Mouse Visium Kidney</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../../_sources/subsections/datasets/mouse_CODEX_spleen.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="mouse-visium-kidney">
<h1>Mouse Visium Kidney<a class="headerlink" href="#mouse-visium-kidney" title="Permalink to this headline">¶</a></h1>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This tutorial was written with <strong>Giotto version 0.3.6.9046</strong>, your version is <strong>1.0.3</strong>. This is a more recent version and results should be reproducible.</p>
</div>
<div class="section" id="install-python-and-r-modules">
<h2>Install Python and R Modules<a class="headerlink" href="#install-python-and-r-modules" title="Permalink to this headline">¶</a></h2>
<p>To run this vignette you need to install all the necessary Python modules. This can be done either (1) manually or (2) via our Installation Tool (from within R).</p>
<ol class="arabic simple">
<li><p><a class="reference internal" href="../../gettingstarted.html#manualinstallation"><span class="std std-ref">Click here for manual installation instructions</span></a></p></li>
<li><p><a class="reference internal" href="../../gettingstarted.html#automaticinstallation"><span class="std std-ref">Click here for our automatic installation tool instructions</span></a></p></li>
</ol>
</div>
<div class="section" id="set-up-giotto">
<h2>Set-Up Giotto<a class="headerlink" href="#set-up-giotto" title="Permalink to this headline">¶</a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">library</span><span class="p">(</span><span class="n">Giotto</span><span class="p">)</span>
</pre></div>
</div>
<div class="section" id="set-a-working-directory">
<h3>Set A Working Directory<a class="headerlink" href="#set-a-working-directory" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#results_folder = &#39;/path/to/directory/&#39;</span>
<span class="n">results_folder</span> <span class="o">=</span> <span class="s1">&#39;/Volumes/Ruben_Seagate/Dropbox (Personal)/Projects/GC_lab/Ruben_Dries/190225_spatial_package/Results/Visium/Brain/201226_results//&#39;</span>
</pre></div>
</div>
</div>
<div class="section" id="set-a-giotto-python-path">
<h3>Set A Giotto Python Path<a class="headerlink" href="#set-a-giotto-python-path" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># set python path to your preferred python version path</span>
<span class="c1"># set python path to NULL if you want to automatically install (only the 1st time) and use the giotto miniconda environment</span>
<span class="n">python_path</span> <span class="o">=</span> <span class="n">NULL</span>
<span class="k">if</span><span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">null</span><span class="p">(</span><span class="n">python_path</span><span class="p">))</span> <span class="p">{</span>
        <span class="n">installGiottoEnvironment</span><span class="p">()</span>
<span class="p">}</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="dataset-explanation">
<h2>Dataset Explanation<a class="headerlink" href="#dataset-explanation" title="Permalink to this headline">¶</a></h2>
<p>The CODEX data to run this tutorial can be found <a class="reference external" href="https://github.com/RubD/spatial-datasets/tree/master/data/2018_codex_spleen">here</a>. Alternatively you can use the <strong>getSpatialDataset</strong> to automatically download this dataset like we do in this example.</p>
<p><a class="reference external" href="https://www.cell.com/cell/pdf/S0092-8674(18)30904-8.pdf">Goltsev et al.</a> created a multiplexed datasets of normal and lupus (MRL/lpr) murine spleens using CODEX technique. The dataset consists of 30 protein markers from 734,101 single cells. In this tutorial, 83,787 cells from sample “BALBc-3” were selected for the analysis.</p>
</div>
<div class="section" id="download-dataset">
<h2>Download Dataset<a class="headerlink" href="#download-dataset" title="Permalink to this headline">¶</a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># download data to working directory</span>
<span class="c1"># use method = &#39;wget&#39; if wget is available. This should be much faster.</span>
<span class="c1"># if you run into authentication issues with wget, then add &quot; extra = &#39;--no-check-certificate&#39; &quot;</span>
<span class="n">getSpatialDataset</span><span class="p">(</span><span class="n">dataset</span> <span class="o">=</span> <span class="s1">&#39;codex_spleen&#39;</span><span class="p">,</span> <span class="n">directory</span> <span class="o">=</span> <span class="n">results_folder</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s1">&#39;wget&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="giotto-global-instructions-and-preparations">
<h2>1. Giotto Global Instructions and Preparations<a class="headerlink" href="#giotto-global-instructions-and-preparations" title="Permalink to this headline">¶</a></h2>
<div class="section" id="optional-set-giotto-instructions">
<h3>1.1 <em>Optional: Set Giotto Instructions</em><a class="headerlink" href="#optional-set-giotto-instructions" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">instrs</span> <span class="o">=</span> <span class="n">createGiottoInstructions</span><span class="p">(</span><span class="n">show_plot</span> <span class="o">=</span> <span class="n">FALSE</span><span class="p">,</span>
                          <span class="n">save_plot</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">,</span>
                          <span class="n">save_dir</span> <span class="o">=</span> <span class="n">results_folder</span><span class="p">,</span>
                          <span class="n">python_path</span> <span class="o">=</span> <span class="n">python_path</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="giotto-object">
<h3>1.2 Giotto Object<a class="headerlink" href="#giotto-object" title="Permalink to this headline">¶</a></h3>
<p>Create Giotto Object from the provided path</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">expr_path</span> <span class="o">=</span> <span class="n">paste0</span><span class="p">(</span><span class="n">results_folder</span><span class="p">,</span> <span class="s2">&quot;codex_BALBc_3_expression.txt.gz&quot;</span><span class="p">)</span>
<span class="n">loc_path</span> <span class="o">=</span> <span class="n">paste0</span><span class="p">(</span><span class="n">results_folder</span><span class="p">,</span> <span class="s2">&quot;codex_BALBc_3_coord.txt&quot;</span><span class="p">)</span>
<span class="n">meta_path</span> <span class="o">=</span> <span class="n">paste0</span><span class="p">(</span><span class="n">results_folder</span><span class="p">,</span> <span class="s2">&quot;codex_BALBc_3_annotation.txt&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="create-giotto-object-and-process-data">
<h2>2. Create Giotto Object and Process Data<a class="headerlink" href="#create-giotto-object-and-process-data" title="Permalink to this headline">¶</a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span># read in data information

# expression info
codex_expression = readExprMatrix(expr_path, transpose = F)
# cell coordinate info
codex_locations = data.table::fread(loc_path)
# metadata
codex_metadata = data.table::fread(meta_path)



## stitch x.y tile coordinates to global coordinates
xtilespan = 1344;
ytilespan = 1008;
# TODO: expand the documentation and input format of stitchTileCoordinates. Probably not enough information for new users.
stitch_file = stitchTileCoordinates(location_file = codex_metadata, Xtilespan = xtilespan, Ytilespan = ytilespan);
codex_locations = stitch_file[,.(Xcoord, Ycoord)]

# create Giotto object
codex_test &lt;- createGiottoObject(raw_exprs = codex_expression,
                         spatial_locs = codex_locations,
                         instructions = instrs,
                         cell_metadata = codex_metadata)

# subset Giotto object
cell_meta = pDataDT(codex_test)
cell_IDs_to_keep = cell_meta[Imaging_phenotype_cell_type != &quot;dirt&quot; &amp; Imaging_phenotype_cell_type != &quot;noid&quot; &amp;                                    Imaging_phenotype_cell_type != &quot;capsule&quot;,]                                                              $cell_ID
codex_test = subsetGiotto(codex_test, cell_ids = cell_IDs_to_keep)

## filter
codex_test &lt;- filterGiotto(gobject = codex_test,
                   expression_threshold = 1,
                   gene_det_in_min_cells = 10,
                   min_det_genes_per_cell = 2,
                   expression_values = c(&#39;raw&#39;),
                   verbose = T)

codex_test &lt;- normalizeGiotto(gobject = codex_test, scalefactor = 6000, verbose = T,
                      log_norm = FALSE,library_size_norm = FALSE,
                      scale_genes = FALSE, scale_cells = TRUE)

## add gene &amp; cell statistics
codex_test &lt;- addStatistics(gobject = codex_test,expression_values = &quot;normalized&quot;)

## adjust expression matrix for technical or known variables
codex_test &lt;- adjustGiottoMatrix(gobject = codex_test,
                         expression_values = c(&#39;normalized&#39;),
                         batch_columns = NULL,
                         covariate_columns = NULL,
                         return_gobject = TRUE,
                         update_slot = c(&#39;custom&#39;))

## visualize
spatPlot(gobject = codex_test,point_size = 0.1,
        coord_fix_ratio = NULL,point_shape = &#39;no_border&#39;,
        save_param = list(save_name = &#39;2_a_spatPlot&#39;))
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/2_a_spatPlot1.png"><img alt="2_a_spatPlot.png" src="../../_images/2_a_spatPlot1.png" style="width: 400px;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">spatPlot</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">codex_test</span><span class="p">,</span> <span class="n">point_size</span> <span class="o">=</span> <span class="mf">0.2</span><span class="p">,</span>
        <span class="n">coord_fix_ratio</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="n">cell_color</span> <span class="o">=</span> <span class="s1">&#39;sample_Xtile_Ytile&#39;</span><span class="p">,</span>
        <span class="n">legend_symbol_size</span> <span class="o">=</span> <span class="mi">3</span><span class="p">,</span><span class="n">legend_text</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span>
        <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;2_b_spatPlot&#39;</span><span class="p">))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/2_b_spatPlot.png"><img alt="2_b_spatPlot.png" src="../../_images/2_b_spatPlot.png" style="width: 400px;" /></a>
</div>
<div class="section" id="dimension-reduction">
<h2>3. Dimension Reduction<a class="headerlink" href="#dimension-reduction" title="Permalink to this headline">¶</a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># use all Abs</span>

<span class="c1"># PCA</span>
<span class="n">codex_test</span> <span class="o">&lt;-</span> <span class="n">runPCA</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">codex_test</span><span class="p">,</span> <span class="n">expression_values</span> <span class="o">=</span> <span class="s1">&#39;normalized&#39;</span><span class="p">,</span> <span class="n">scale_unit</span> <span class="o">=</span> <span class="n">T</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s2">&quot;factominer&quot;</span><span class="p">)</span>
<span class="n">signPCA</span><span class="p">(</span><span class="n">codex_test</span><span class="p">,</span> <span class="n">scale_unit</span> <span class="o">=</span> <span class="n">T</span><span class="p">,</span> <span class="n">scree_ylim</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">),</span>
                <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;3_a_spatPlot&#39;</span><span class="p">))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/3_a_spatPlot.png"><img alt="3_a_spatPlot.png" src="../../_images/3_a_spatPlot.png" style="width: 400px;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">plotPCA</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">codex_test</span><span class="p">,</span> <span class="n">point_shape</span> <span class="o">=</span> <span class="s1">&#39;no_border&#39;</span><span class="p">,</span> <span class="n">point_size</span> <span class="o">=</span> <span class="mf">0.2</span><span class="p">,</span>
                <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;3_b_PCA&#39;</span><span class="p">))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/3_b_PCA1.png"><img alt="3_b_PCA.png" src="../../_images/3_b_PCA1.png" style="width: 400px;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># UMAP</span>
<span class="n">codex_test</span> <span class="o">&lt;-</span> <span class="n">runUMAP</span><span class="p">(</span><span class="n">codex_test</span><span class="p">,</span> <span class="n">dimensions_to_use</span> <span class="o">=</span> <span class="mi">1</span><span class="p">:</span><span class="mi">14</span><span class="p">,</span> <span class="n">n_components</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">n_threads</span> <span class="o">=</span> <span class="mi">12</span><span class="p">)</span>
<span class="n">plotUMAP</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">codex_test</span><span class="p">,</span> <span class="n">point_shape</span> <span class="o">=</span> <span class="s1">&#39;no_border&#39;</span><span class="p">,</span> <span class="n">point_size</span> <span class="o">=</span> <span class="mf">0.2</span><span class="p">,</span>
        <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;3_c_UMAP&#39;</span><span class="p">))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/3_c_UMAP1.png"><img alt="3_c_UMAP.png" src="../../_images/3_c_UMAP1.png" style="width: 400px;" /></a>
</div>
<div class="section" id="cluster">
<h2>4. Cluster<a class="headerlink" href="#cluster" title="Permalink to this headline">¶</a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>## sNN network (default)
codex_test &lt;- createNearestNetwork(gobject = codex_test, dimensions_to_use = 1:14, k = 20)

## 0.1 resolution
codex_test &lt;- doLeidenCluster(gobject = codex_test, resolution = 0.5, n_iterations = 100, name =        &#39;leiden&#39;,python_path = python_path)

codex_metadata = pDataDT(codex_test)
leiden_colors = Giotto:::getDistinctColors(length(unique(codex_metadata$leiden)))
names(leiden_colors) = unique(codex_metadata$leiden)

plotUMAP(gobject = codex_test,
        cell_color = &#39;leiden&#39;, point_shape = &#39;no_border&#39;, point_size = 0.2, cell_color_code = leiden_colors,
        save_param = list(save_name = &#39;4_a_UMAP&#39;))
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/4_a_UMAP2.png"><img alt="4_a_UMAP.png" src="../../_images/4_a_UMAP2.png" style="width: 400px;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">spatPlot</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">codex_test</span><span class="p">,</span> <span class="n">cell_color</span> <span class="o">=</span> <span class="s1">&#39;leiden&#39;</span><span class="p">,</span> <span class="n">point_shape</span> <span class="o">=</span> <span class="s1">&#39;no_border&#39;</span><span class="p">,</span> <span class="n">point_size</span> <span class="o">=</span> <span class="mf">0.2</span><span class="p">,</span>
        <span class="n">cell_color_code</span> <span class="o">=</span> <span class="n">leiden_colors</span><span class="p">,</span> <span class="n">coord_fix_ratio</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span><span class="n">label_size</span> <span class="o">=</span><span class="mi">2</span><span class="p">,</span>
        <span class="n">legend_text</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span><span class="n">legend_symbol_size</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span>
        <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;4_b_spatplot&#39;</span><span class="p">))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/4_b_spatplot.png"><img alt="4_b_spatplot.png" src="../../_images/4_b_spatplot.png" style="width: 400px;" /></a>
</div>
<div class="section" id="co-visualize">
<h2>5. Co-Visualize<a class="headerlink" href="#co-visualize" title="Permalink to this headline">¶</a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">spatDimPlot2D</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">codex_test</span><span class="p">,</span> <span class="n">cell_color</span> <span class="o">=</span> <span class="s1">&#39;leiden&#39;</span><span class="p">,</span> <span class="n">spat_point_shape</span> <span class="o">=</span> <span class="s1">&#39;no_border&#39;</span><span class="p">,</span>
      <span class="n">spat_point_size</span> <span class="o">=</span> <span class="mf">0.2</span><span class="p">,</span> <span class="n">dim_point_shape</span> <span class="o">=</span> <span class="s1">&#39;no_border&#39;</span><span class="p">,</span> <span class="n">dim_point_size</span> <span class="o">=</span> <span class="mf">0.2</span><span class="p">,</span>
      <span class="n">cell_color_code</span> <span class="o">=</span> <span class="n">leiden_colors</span><span class="p">,</span><span class="n">plot_alignment</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s2">&quot;horizontal&quot;</span><span class="p">),</span>
      <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;5_a_spatdimplot&#39;</span><span class="p">))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/5_a_spatdimplot1.png"><img alt="5_a_spatdimplot.png" src="../../_images/5_a_spatdimplot1.png" style="width: 400px;" /></a>
</div>
<div class="section" id="differential-expression">
<h2>6. Differential Expression<a class="headerlink" href="#differential-expression" title="Permalink to this headline">¶</a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># resolution 0.5</span>
<span class="n">cluster_column</span> <span class="o">=</span> <span class="s1">&#39;leiden&#39;</span>
<span class="n">markers_scran</span> <span class="o">=</span> <span class="n">findMarkers_one_vs_all</span><span class="p">(</span><span class="n">gobject</span><span class="o">=</span><span class="n">codex_test</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s2">&quot;scran&quot;</span><span class="p">,</span>
                               <span class="n">expression_values</span><span class="o">=</span><span class="s2">&quot;norm&quot;</span><span class="p">,</span> <span class="n">cluster_column</span><span class="o">=</span><span class="n">cluster_column</span><span class="p">,</span> <span class="n">min_genes</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="n">markergenes_scran</span> <span class="o">=</span> <span class="n">unique</span><span class="p">(</span><span class="n">markers_scran</span><span class="p">[,</span> <span class="n">head</span><span class="p">(</span><span class="o">.</span><span class="n">SD</span><span class="p">,</span> <span class="mi">5</span><span class="p">),</span> <span class="n">by</span><span class="o">=</span><span class="s2">&quot;cluster&quot;</span><span class="p">][[</span><span class="s2">&quot;genes&quot;</span><span class="p">]])</span>

<span class="n">plotMetaDataHeatmap</span><span class="p">(</span><span class="n">codex_test</span><span class="p">,</span> <span class="n">expression_values</span> <span class="o">=</span> <span class="s2">&quot;norm&quot;</span><span class="p">,</span> <span class="n">metadata_cols</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="n">cluster_column</span><span class="p">),</span>
            <span class="n">selected_genes</span> <span class="o">=</span> <span class="n">markergenes_scran</span><span class="p">,</span>
            <span class="n">y_text_size</span> <span class="o">=</span> <span class="mi">8</span><span class="p">,</span> <span class="n">show_values</span> <span class="o">=</span> <span class="s1">&#39;zscores_rescaled&#39;</span><span class="p">,</span>
            <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;6_a_metaheatmap&#39;</span><span class="p">))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/6_a_metaheatmap1.png"><img alt="6_a_metaheatmap.png" src="../../_images/6_a_metaheatmap1.png" style="width: 400px;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>topgenes_scran = markers_scran[, head(.SD, 1), by = &#39;cluster&#39;]$genes
violinPlot(codex_test, genes = unique(topgenes_scran)[1:8], cluster_column = cluster_column,
   strip_text = 8, strip_position = &#39;right&#39;,
   save_param = list(save_name = &#39;6_b_violinplot&#39;))
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/6_b_violinplot.png"><img alt="6_b_violinplot.png" src="../../_images/6_b_violinplot.png" style="width: 400px;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># gini</span>
<span class="n">markers_gini</span> <span class="o">=</span> <span class="n">findMarkers_one_vs_all</span><span class="p">(</span><span class="n">gobject</span><span class="o">=</span><span class="n">codex_test</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s2">&quot;gini&quot;</span><span class="p">,</span> <span class="n">expression_values</span><span class="o">=</span><span class="s2">&quot;norm&quot;</span><span class="p">,</span>
                              <span class="n">cluster_column</span><span class="o">=</span><span class="n">cluster_column</span><span class="p">,</span> <span class="n">min_genes</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
<span class="n">markergenes_gini</span> <span class="o">=</span> <span class="n">unique</span><span class="p">(</span><span class="n">markers_gini</span><span class="p">[,</span> <span class="n">head</span><span class="p">(</span><span class="o">.</span><span class="n">SD</span><span class="p">,</span> <span class="mi">5</span><span class="p">),</span> <span class="n">by</span><span class="o">=</span><span class="s2">&quot;cluster&quot;</span><span class="p">][[</span><span class="s2">&quot;genes&quot;</span><span class="p">]])</span>
<span class="n">plotMetaDataHeatmap</span><span class="p">(</span><span class="n">codex_test</span><span class="p">,</span> <span class="n">expression_values</span> <span class="o">=</span> <span class="s2">&quot;norm&quot;</span><span class="p">,</span>
            <span class="n">metadata_cols</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="n">cluster_column</span><span class="p">),</span> <span class="n">selected_genes</span> <span class="o">=</span> <span class="n">markergenes_gini</span><span class="p">,</span>
            <span class="n">show_values</span> <span class="o">=</span> <span class="s1">&#39;zscores_rescaled&#39;</span><span class="p">,</span>
            <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;6_c_metaheatmap&#39;</span><span class="p">))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/6_c_metaheatmap.png"><img alt="6_c_metaheatmap.png" src="../../_images/6_c_metaheatmap.png" style="width: 400px;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>topgenes_gini = markers_gini[, head(.SD, 1), by = &#39;cluster&#39;]$genes
violinPlot(codex_test, genes = unique(topgenes_gini), cluster_column = cluster_column,
        strip_text = 8, strip_position = &#39;right&#39;,
        save_param = list(save_name = &#39;6_d_violinplot&#39;))
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/6_d_violinplot.png"><img alt="6_d_violinplot.png" src="../../_images/6_d_violinplot.png" style="width: 400px;" /></a>
</div>
<div class="section" id="cell-type-annotation">
<h2>7. Cell-Type Annotation<a class="headerlink" href="#cell-type-annotation" title="Permalink to this headline">¶</a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">clusters_cell_types</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s1">&#39;erythroblasts-F4/80(+) mphs&#39;</span><span class="p">,</span><span class="s1">&#39;B cells&#39;</span><span class="p">,</span><span class="s1">&#39;CD8(+) T cells&#39;</span><span class="p">,</span>
                <span class="s1">&#39;CD4(+) T cells&#39;</span><span class="p">,</span> <span class="s1">&#39;B cells&#39;</span><span class="p">,</span><span class="s1">&#39;CD11c(+)MHCII(+) cells&#39;</span><span class="p">,</span>
                <span class="s1">&#39;CD4(+) T cells&#39;</span><span class="p">,</span><span class="s1">&#39;Ter119(+)&#39;</span><span class="p">,</span> <span class="s1">&#39;marginal zone mphs&#39;</span><span class="p">,</span>
                <span class="s1">&#39;CD31(+)ERTR7(+)&#39;</span><span class="p">,</span> <span class="s1">&#39;FDCs&#39;</span><span class="p">,</span> <span class="s1">&#39;B220(+) DN T cells&#39;</span><span class="p">,</span>
                <span class="s1">&#39;CD3(+) other markers&#39;</span><span class="p">,</span><span class="s1">&#39;NK cells&#39;</span><span class="p">,</span><span class="s1">&#39;granulocytes&#39;</span><span class="p">,</span>
                <span class="s1">&#39;plasma cells&#39;</span><span class="p">,</span><span class="s1">&#39;ambiguous&#39;</span><span class="p">,</span><span class="s1">&#39;CD44(+)CD1632(+)Ly6C(+) cells&#39;</span><span class="p">)</span>

<span class="n">names</span><span class="p">(</span><span class="n">clusters_cell_types</span><span class="p">)</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="mi">1</span><span class="p">:</span><span class="mi">18</span><span class="p">)</span>
<span class="n">codex_test</span> <span class="o">=</span> <span class="n">annotateGiotto</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">codex_test</span><span class="p">,</span> <span class="n">annotation_vector</span> <span class="o">=</span> <span class="n">clusters_cell_types</span><span class="p">,</span>
                    <span class="n">cluster_column</span> <span class="o">=</span> <span class="s1">&#39;leiden&#39;</span><span class="p">,</span> <span class="n">name</span> <span class="o">=</span> <span class="s1">&#39;cell_types&#39;</span><span class="p">)</span>

<span class="n">plotMetaDataHeatmap</span><span class="p">(</span><span class="n">codex_test</span><span class="p">,</span> <span class="n">expression_values</span> <span class="o">=</span> <span class="s1">&#39;scaled&#39;</span><span class="p">,</span>
            <span class="n">metadata_cols</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s1">&#39;cell_types&#39;</span><span class="p">),</span><span class="n">y_text_size</span> <span class="o">=</span> <span class="mi">6</span><span class="p">,</span>
            <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;7_a_metaheatmap&#39;</span><span class="p">))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/7_a_metaheatmap1.png"><img alt="7_a_metaheatmap.png" src="../../_images/7_a_metaheatmap1.png" style="width: 400px;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span># create consistent color code
mynames = unique(pDataDT(codex_test)$cell_types)
mycolorcode = Giotto:::getDistinctColors(n = length(mynames))
names(mycolorcode) = mynames

plotUMAP(gobject = codex_test, cell_color = &#39;cell_types&#39;,point_shape = &#39;no_border&#39;,   point_size = 0.2,
        cell_color_code = mycolorcode,
        show_center_label = F,
        label_size =2,
        legend_text = 5,
        legend_symbol_size = 2,
        save_param = list(save_name = &#39;7_b_umap&#39;))
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/7_b_umap.png"><img alt="7_b_umap.png" src="../../_images/7_b_umap.png" style="width: 400px;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">spatPlot</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">codex_test</span><span class="p">,</span> <span class="n">cell_color</span> <span class="o">=</span> <span class="s1">&#39;cell_types&#39;</span><span class="p">,</span> <span class="n">point_shape</span> <span class="o">=</span> <span class="s1">&#39;no_border&#39;</span><span class="p">,</span> <span class="n">point_size</span> <span class="o">=</span> <span class="mf">0.2</span><span class="p">,</span>
        <span class="n">cell_color_code</span> <span class="o">=</span> <span class="n">mycolorcode</span><span class="p">,</span>
        <span class="n">coord_fix_ratio</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
        <span class="n">label_size</span> <span class="o">=</span><span class="mi">2</span><span class="p">,</span>
                <span class="n">legend_text</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span>
        <span class="n">legend_symbol_size</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span>
        <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;7_c_spatplot&#39;</span><span class="p">))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/7_c_spatplot1.png"><img alt="7_c_spatplot.png" src="../../_images/7_c_spatplot1.png" style="width: 400px;" /></a>
</div>
<div class="section" id="cell-type-visualization-and-gene-expression-of-selected-zones">
<h2>8. Cell-Type Visualization and Gene Expression of Selected Zones<a class="headerlink" href="#cell-type-visualization-and-gene-expression-of-selected-zones" title="Permalink to this headline">¶</a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>cell_metadata = pDataDT(codex_test)
subset_cell_ids = cell_metadata[sample_Xtile_Ytile==&quot;BALBc-3_X04_Y08&quot;]$cell_ID

codex_test_zone1 = subsetGiotto(codex_test, cell_ids = subset_cell_ids)

plotUMAP(gobject = codex_test_zone1,
        cell_color = &#39;cell_types&#39;, point_shape = &#39;no_border&#39;,   point_size = 1,
        cell_color_code = mycolorcode,
        show_center_label = F,
        label_size =2,
                legend_text = 5,
        legend_symbol_size = 2,
        save_param = list(save_name = &#39;8_a_umap&#39;))
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/8_a_umap.png"><img alt="8_a_umap.png" src="../../_images/8_a_umap.png" style="width: 400px;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">spatPlot</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">codex_test_zone1</span><span class="p">,</span>
        <span class="n">cell_color</span> <span class="o">=</span> <span class="s1">&#39;cell_types&#39;</span><span class="p">,</span> <span class="n">point_shape</span> <span class="o">=</span> <span class="s1">&#39;no_border&#39;</span><span class="p">,</span> <span class="n">point_size</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
        <span class="n">cell_color_code</span> <span class="o">=</span> <span class="n">mycolorcode</span><span class="p">,</span>
        <span class="n">coord_fix_ratio</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
        <span class="n">label_size</span> <span class="o">=</span><span class="mi">2</span><span class="p">,</span>
        <span class="n">legend_text</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span>
        <span class="n">legend_symbol_size</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span>
        <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;8_b_spatplot&#39;</span><span class="p">))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/8_b_spatplot1.png"><img alt="8_b_spatplot.png" src="../../_images/8_b_spatplot1.png" style="width: 400px;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">spatDimGenePlot</span><span class="p">(</span><span class="n">codex_test_zone1</span><span class="p">,</span>
        <span class="n">expression_values</span> <span class="o">=</span> <span class="s1">&#39;scaled&#39;</span><span class="p">,</span>
        <span class="n">genes</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s2">&quot;CD8a&quot;</span><span class="p">,</span><span class="s2">&quot;CD19&quot;</span><span class="p">),</span>
        <span class="n">spat_point_shape</span> <span class="o">=</span> <span class="s1">&#39;no_border&#39;</span><span class="p">,</span>
        <span class="n">dim_point_shape</span> <span class="o">=</span> <span class="s1">&#39;no_border&#39;</span><span class="p">,</span>
        <span class="n">cell_color_gradient</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s2">&quot;darkblue&quot;</span><span class="p">,</span> <span class="s2">&quot;white&quot;</span><span class="p">,</span> <span class="s2">&quot;red&quot;</span><span class="p">),</span>
        <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;8_c_spatdimplot&#39;</span><span class="p">))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/8_c_spatdimplot.png"><img alt="8_c_spatdimplot.png" src="../../_images/8_c_spatdimplot.png" style="width: 400px;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>cell_metadata = pDataDT(codex_test)
subset_cell_ids = cell_metadata[sample_Xtile_Ytile==&quot;BALBc-3_X04_Y03&quot;]$cell_ID

codex_test_zone2 = subsetGiotto(codex_test, cell_ids = subset_cell_ids)

plotUMAP(gobject = codex_test_zone2, cell_color = &#39;cell_types&#39;,point_shape = &#39;no_border&#39;,   point_size = 1,
        cell_color_code = mycolorcode,
        show_center_label = F,
        label_size =2,
        legend_text = 5,
        legend_symbol_size = 2,
        save_param = list(save_name = &#39;8_d_umap&#39;))
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/8_d_umap.png"><img alt="8_d_umap.png" src="../../_images/8_d_umap.png" style="width: 400px;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">spatPlot</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">codex_test_zone2</span><span class="p">,</span> <span class="n">cell_color</span> <span class="o">=</span> <span class="s1">&#39;cell_types&#39;</span><span class="p">,</span> <span class="n">point_shape</span> <span class="o">=</span> <span class="s1">&#39;no_border&#39;</span><span class="p">,</span> <span class="n">point_size</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
        <span class="n">cell_color_code</span> <span class="o">=</span> <span class="n">mycolorcode</span><span class="p">,</span>
        <span class="n">coord_fix_ratio</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
        <span class="n">label_size</span> <span class="o">=</span><span class="mi">2</span><span class="p">,</span>
        <span class="n">legend_text</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span>
        <span class="n">legend_symbol_size</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span>
                <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;8_e_spatPlot&#39;</span><span class="p">))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/8_e_spatPlot.png"><img alt="8_e_spatPlot.png" src="../../_images/8_e_spatPlot.png" style="width: 400px;" /></a>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">spatDimGenePlot</span><span class="p">(</span><span class="n">codex_test_zone2</span><span class="p">,</span>
        <span class="n">expression_values</span> <span class="o">=</span> <span class="s1">&#39;scaled&#39;</span><span class="p">,</span>
        <span class="n">genes</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s2">&quot;CD4&quot;</span><span class="p">,</span> <span class="s2">&quot;CD106&quot;</span><span class="p">),</span>
        <span class="n">spat_point_shape</span> <span class="o">=</span> <span class="s1">&#39;no_border&#39;</span><span class="p">,</span>
        <span class="n">dim_point_shape</span> <span class="o">=</span> <span class="s1">&#39;no_border&#39;</span><span class="p">,</span>
        <span class="n">cell_color_gradient</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s2">&quot;darkblue&quot;</span><span class="p">,</span> <span class="s2">&quot;white&quot;</span><span class="p">,</span> <span class="s2">&quot;red&quot;</span><span class="p">),</span>
        <span class="n">save_param</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">save_name</span> <span class="o">=</span> <span class="s1">&#39;8_f_spatdimgeneplot&#39;</span><span class="p">))</span>
</pre></div>
</div>
<a class="reference internal image-reference" href="../../_images/8_f_spatdimgeneplot.png"><img alt="8_f_spatdimgeneplot.png" src="../../_images/8_f_spatdimgeneplot.png" style="width: 400px;" /></a>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="osmFISH_mouse_SS_cortex.html" class="btn btn-neutral float-right" title="Mouse osmFISH SS Cortex" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="mouse_visium_kidney.html" class="btn btn-neutral float-left" title="Mouse Visium Kidney" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2021, Ruben Dries, Qian Zhu, Huipeng Li, Rui Dong, Guo-Cheng Yuan.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>