

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>mini Visium &mdash; Giotto 1.0 documentation</title>
  

  
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/togglebutton.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/css/theme_edits.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/panels-bootstrap.5fd3999ee7762ccc51105388f4a9d115.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" type="text/css" />
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script src="../../_static/togglebutton.js"></script>
        <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    
    <script type="text/javascript" src="../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="mini3D STARmap" href="mini_3D_STARmap.html" />
    <link rel="prev" title="mini seqFISH" href="mini_seqFISH.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../index.html">
          

          
            
            <img src="../../_static/GiottoLogo.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption" role="heading"><span class="caption-text">General</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../gettingstarted.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../giottoworkflowanalyses.html">Giotto Workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../datasets.html">Dataset Information</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../datasets.html#examples">Examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../documentation.html">Giotto Functions</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../tipsandtricks.html">Tips and Tricks</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../trygiotto.html">Try Giotto</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Supplemental Information</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../errorsfaqsandtips.html">Errors and FAQS</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../additionalinformation.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/RubD/Giotto/">Browse Source Code</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/RubD/Giotto/issues">Report a Bug</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../contribute.html">Contribute</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">Giotto</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../../datasets.html">Dataset Information</a> &raquo;</li>
        
      <li>mini Visium</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../../_sources/subsections/datasets/mini_visium.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="mini-visium">
<h1>mini Visium<a class="headerlink" href="#mini-visium" title="Permalink to this headline">¶</a></h1>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">library</span><span class="p">(</span><span class="n">Giotto</span><span class="p">)</span>
</pre></div>
</div>
<div class="section" id="install-python-modules">
<h2>Install Python Modules<a class="headerlink" href="#install-python-modules" title="Permalink to this headline">¶</a></h2>
<p>To run this vignette you need to install all the necessary Python modules. This can be done either (1) manually or (2) via our Installation Tool (from within R).</p>
<ol class="arabic simple">
<li><p><a class="reference internal" href="../../gettingstarted.html#manualinstallation"><span class="std std-ref">Click here for manual installation instructions</span></a></p></li>
<li><p><a class="reference internal" href="../../gettingstarted.html#automaticinstallation"><span class="std std-ref">Click here for our automatic installation tool instructions</span></a></p></li>
</ol>
</div>
<div class="section" id="optional-set-giotto-instructions">
<h2><em>Optional: Set Giotto Instructions</em><a class="headerlink" href="#optional-set-giotto-instructions" title="Permalink to this headline">¶</a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># to automatically save figures in save_dir set save_plot to TRUE</span>
<span class="n">temp_dir</span> <span class="o">=</span> <span class="n">getwd</span><span class="p">()</span>
<span class="n">temp_dir</span> <span class="o">=</span> <span class="s1">&#39;~/Temp/&#39;</span>
<span class="n">myinstructions</span> <span class="o">=</span> <span class="n">createGiottoInstructions</span><span class="p">(</span><span class="n">save_dir</span> <span class="o">=</span> <span class="n">temp_dir</span><span class="p">,</span>
                                      <span class="n">save_plot</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">,</span>
                                      <span class="n">show_plot</span> <span class="o">=</span> <span class="n">FALSE</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="working-with-giotto-object">
<h2>1. Working with Giotto Object<a class="headerlink" href="#working-with-giotto-object" title="Permalink to this headline">¶</a></h2>
<div class="section" id="create-giotto-object">
<h3>1.1 Create Giotto Object<a class="headerlink" href="#create-giotto-object" title="Permalink to this headline">¶</a></h3>
<div class="section" id="minimum-requirements">
<h4>Minimum Requirements:<a class="headerlink" href="#minimum-requirements" title="Permalink to this headline">¶</a></h4>
<ul class="simple">
<li><p>Matrix with expression information (or path to)</p></li>
<li><p>x,y(,z) coordinates for cells or spots (or path to)</p></li>
</ul>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># giotto object</span>
    <span class="n">expr_path</span> <span class="o">=</span> <span class="n">system</span><span class="o">.</span><span class="n">file</span><span class="p">(</span><span class="s2">&quot;extdata&quot;</span><span class="p">,</span> <span class="s2">&quot;visium_DG_expr.txt.gz&quot;</span><span class="p">,</span> <span class="n">package</span> <span class="o">=</span> <span class="s1">&#39;Giotto&#39;</span><span class="p">)</span>
    <span class="n">loc_path</span> <span class="o">=</span> <span class="n">system</span><span class="o">.</span><span class="n">file</span><span class="p">(</span><span class="s2">&quot;extdata&quot;</span><span class="p">,</span> <span class="s2">&quot;visium_DG_locs.txt&quot;</span><span class="p">,</span> <span class="n">package</span> <span class="o">=</span> <span class="s1">&#39;Giotto&#39;</span><span class="p">)</span>
    <span class="n">mini_visium</span> <span class="o">&lt;-</span> <span class="n">createGiottoObject</span><span class="p">(</span><span class="n">raw_exprs</span> <span class="o">=</span> <span class="n">expr_path</span><span class="p">,</span>
                                    <span class="n">spatial_locs</span> <span class="o">=</span> <span class="n">loc_path</span><span class="p">,</span>
                                    <span class="n">instructions</span> <span class="o">=</span> <span class="n">myinstructions</span><span class="p">)</span>
</pre></div>
</div>
<p>How to work with Giotto instructions that are part of your Giotto object:</p>
<ul class="simple">
<li><p>Show the instructions associated with your Giotto object with <strong>showGiottoInstructions</strong></p></li>
<li><p>Change one or more instructions with <strong>changeGiottoInstructions</strong></p></li>
<li><p>Replace all instructions at once with <strong>replaceGiottoInstructions</strong></p></li>
<li><p>Read or get a specific giotto instruction with <strong>readGiottoInstructions</strong></p></li>
</ul>
<p><em>Note: The python path can only be set once in an R session. See the **reticulate package*</em> for more information.*</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># show instructions associated with giotto object (mini_visium)</span>
<span class="n">showGiottoInstructions</span><span class="p">(</span><span class="n">mini_visium</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="add-image">
<h3>1.2 Add Image<a class="headerlink" href="#add-image" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">## 1. read image</span>
<span class="n">png_path</span> <span class="o">=</span> <span class="n">system</span><span class="o">.</span><span class="n">file</span><span class="p">(</span><span class="s2">&quot;extdata&quot;</span><span class="p">,</span> <span class="s2">&quot;deg_image.png&quot;</span><span class="p">,</span> <span class="n">package</span> <span class="o">=</span> <span class="s1">&#39;Giotto&#39;</span><span class="p">)</span>
<span class="n">mg_img</span> <span class="o">=</span> <span class="n">magick</span><span class="p">::</span><span class="n">image_read</span><span class="p">(</span><span class="n">png_path</span><span class="p">)</span>

<span class="c1">## 2. test and modify image alignment</span>
<span class="n">mypl</span> <span class="o">=</span> <span class="n">spatPlot</span><span class="p">(</span><span class="n">mini_visium</span><span class="p">,</span> <span class="n">return_plot</span> <span class="o">=</span> <span class="n">T</span><span class="p">,</span> <span class="n">point_alpha</span> <span class="o">=</span> <span class="mf">0.8</span><span class="p">)</span>
<span class="n">orig_png</span> <span class="o">=</span> <span class="n">createGiottoImage</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">mini_visium</span><span class="p">,</span> <span class="n">mg_object</span> <span class="o">=</span> <span class="n">mg_img</span><span class="p">,</span> <span class="n">name</span> <span class="o">=</span> <span class="s1">&#39;image&#39;</span><span class="p">,</span>
                        <span class="n">xmax_adj</span> <span class="o">=</span> <span class="mi">450</span><span class="p">,</span> <span class="n">xmin_adj</span> <span class="o">=</span> <span class="mi">550</span><span class="p">,</span>
                        <span class="n">ymax_adj</span> <span class="o">=</span> <span class="mi">200</span><span class="p">,</span> <span class="n">ymin_adj</span> <span class="o">=</span> <span class="mi">200</span><span class="p">)</span>
<span class="n">mypl_image</span> <span class="o">=</span> <span class="n">addGiottoImageToSpatPlot</span><span class="p">(</span><span class="n">mypl</span><span class="p">,</span> <span class="n">orig_png</span><span class="p">)</span>
<span class="n">mypl_image</span>

<span class="c1">## 3. add images to Giotto object ##</span>
<span class="n">image_list</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">orig_png</span><span class="p">)</span>
<span class="n">mini_visium</span> <span class="o">=</span> <span class="n">addGiottoImage</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">mini_visium</span><span class="p">,</span>
                        <span class="n">images</span> <span class="o">=</span> <span class="n">image_list</span><span class="p">)</span>
<span class="n">showGiottoImageNames</span><span class="p">(</span><span class="n">mini_visium</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="processing-steps">
<h2>2. Processing Steps<a class="headerlink" href="#processing-steps" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Filter genes and cells based on detection frequencies</p></li>
<li><p>Normalize expression matrix (log transformation, scaling factor and/or z-scores)</p></li>
<li><p>Add cell and gene statistics (optional)</p></li>
<li><p>Adjust expression matrix for technical covariates or batches (optional). These results will be stored in the custom slot.</p></li>
</ul>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># explore gene and cell distribution</span>
<span class="n">filterDistributions</span><span class="p">(</span><span class="n">mini_visium</span><span class="p">,</span> <span class="n">detection</span> <span class="o">=</span> <span class="s1">&#39;genes&#39;</span><span class="p">)</span>
<span class="n">filterDistributions</span><span class="p">(</span><span class="n">mini_visium</span><span class="p">,</span> <span class="n">detection</span> <span class="o">=</span> <span class="s1">&#39;cells&#39;</span><span class="p">)</span>
<span class="n">filterCombinations</span><span class="p">(</span><span class="n">mini_visium</span><span class="p">,</span>
                <span class="n">expression_thresholds</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="mi">1</span><span class="p">),</span>
                <span class="n">gene_det_in_min_cells</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="mi">20</span><span class="p">,</span> <span class="mi">20</span><span class="p">,</span> <span class="mi">50</span><span class="p">,</span> <span class="mi">50</span><span class="p">),</span>
                <span class="n">min_det_genes_per_cell</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="mi">100</span><span class="p">,</span> <span class="mi">200</span><span class="p">,</span> <span class="mi">100</span><span class="p">,</span> <span class="mi">200</span><span class="p">))</span>

<span class="c1"># filter and normalize</span>
<span class="n">mini_visium</span> <span class="o">&lt;-</span> <span class="n">filterGiotto</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">mini_visium</span><span class="p">,</span>
                        <span class="n">expression_threshold</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
                        <span class="n">gene_det_in_min_cells</span> <span class="o">=</span> <span class="mi">50</span><span class="p">,</span>
                        <span class="n">min_det_genes_per_cell</span> <span class="o">=</span> <span class="mi">100</span><span class="p">,</span>
                        <span class="n">expression_values</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s1">&#39;raw&#39;</span><span class="p">),</span>
                        <span class="n">verbose</span> <span class="o">=</span> <span class="n">T</span><span class="p">)</span>
<span class="n">mini_visium</span> <span class="o">&lt;-</span> <span class="n">normalizeGiotto</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">mini_visium</span><span class="p">,</span> <span class="n">scalefactor</span> <span class="o">=</span> <span class="mi">6000</span><span class="p">,</span> <span class="n">verbose</span> <span class="o">=</span> <span class="n">T</span><span class="p">)</span>
<span class="n">mini_visium</span> <span class="o">&lt;-</span> <span class="n">addStatistics</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">mini_visium</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="dimension-reduction">
<h2>3. Dimension Reduction<a class="headerlink" href="#dimension-reduction" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Identify highly variable genes (HVG)</p></li>
<li><p>Perform PCA</p></li>
<li><p>Identify number of significant prinicipal components (PCs)</p></li>
<li><p>Run UMAP and/or TSNE on PCs (or directly on matrix)</p></li>
</ul>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">mini_visium</span> <span class="o">&lt;-</span> <span class="n">calculateHVG</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">mini_visium</span><span class="p">)</span>

<span class="n">mini_visium</span> <span class="o">&lt;-</span> <span class="n">runPCA</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">mini_visium</span><span class="p">)</span>
<span class="n">screePlot</span><span class="p">(</span><span class="n">mini_visium</span><span class="p">,</span> <span class="n">ncp</span> <span class="o">=</span> <span class="mi">30</span><span class="p">)</span>
<span class="n">plotPCA</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">mini_visium</span><span class="p">)</span>

<span class="n">mini_visium</span> <span class="o">&lt;-</span> <span class="n">runUMAP</span><span class="p">(</span><span class="n">mini_visium</span><span class="p">,</span> <span class="n">dimensions_to_use</span> <span class="o">=</span> <span class="mi">1</span><span class="p">:</span><span class="mi">10</span><span class="p">)</span>
<span class="n">plotUMAP</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">mini_visium</span><span class="p">)</span>
<span class="n">mini_visium</span> <span class="o">&lt;-</span> <span class="n">runtSNE</span><span class="p">(</span><span class="n">mini_visium</span><span class="p">,</span> <span class="n">dimensions_to_use</span> <span class="o">=</span> <span class="mi">1</span><span class="p">:</span><span class="mi">10</span><span class="p">)</span>
<span class="n">plotTSNE</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">mini_visium</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="clustering">
<h2>4. Clustering<a class="headerlink" href="#clustering" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Create a shared (default) nearest network in PCA space (or directly on matrix)</p></li>
<li><p>Cluster on nearest network with Leiden or Louvan (kmeans and hclust are alternatives)</p></li>
</ul>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">mini_visium</span> <span class="o">&lt;-</span> <span class="n">createNearestNetwork</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">mini_visium</span><span class="p">,</span> <span class="n">dimensions_to_use</span> <span class="o">=</span> <span class="mi">1</span><span class="p">:</span><span class="mi">10</span><span class="p">,</span> <span class="n">k</span> <span class="o">=</span> <span class="mi">15</span><span class="p">)</span>
<span class="n">mini_visium</span> <span class="o">&lt;-</span> <span class="n">doLeidenCluster</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">mini_visium</span><span class="p">,</span> <span class="n">resolution</span> <span class="o">=</span> <span class="mf">0.4</span><span class="p">,</span> <span class="n">n_iterations</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">)</span>

<span class="n">pDataDT</span><span class="p">(</span><span class="n">mini_visium</span><span class="p">)</span>

<span class="c1"># visualize UMAP cluster results</span>
<span class="n">plotUMAP</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">mini_visium</span><span class="p">,</span> <span class="n">cell_color</span> <span class="o">=</span> <span class="s1">&#39;leiden_clus&#39;</span><span class="p">,</span> <span class="n">show_NN_network</span> <span class="o">=</span> <span class="n">T</span><span class="p">,</span> <span class="n">point_size</span> <span class="o">=</span> <span class="mf">2.5</span><span class="p">)</span>

<span class="c1"># visualize UMAP and spatial results</span>
<span class="n">spatDimPlot</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">mini_visium</span><span class="p">,</span> <span class="n">cell_color</span> <span class="o">=</span> <span class="s1">&#39;leiden_clus&#39;</span><span class="p">,</span> <span class="n">spat_point_shape</span> <span class="o">=</span> <span class="s1">&#39;voronoi&#39;</span><span class="p">)</span>

<span class="c1"># heatmap and dendrogram</span>
<span class="n">showClusterHeatmap</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">mini_visium</span><span class="p">,</span> <span class="n">cluster_column</span> <span class="o">=</span> <span class="s1">&#39;leiden_clus&#39;</span><span class="p">)</span>
<span class="n">showClusterDendrogram</span><span class="p">(</span><span class="n">mini_visium</span><span class="p">,</span> <span class="n">h</span> <span class="o">=</span> <span class="mf">0.5</span><span class="p">,</span> <span class="n">rotate</span> <span class="o">=</span> <span class="n">T</span><span class="p">,</span> <span class="n">cluster_column</span> <span class="o">=</span> <span class="s1">&#39;leiden_clus&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="differential-expression">
<h2>5. Differential Expression<a class="headerlink" href="#differential-expression" title="Permalink to this headline">¶</a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>scran_markers = findMarkers_one_vs_all(gobject = mini_visium,
                                method = &#39;scran&#39;,
                                expression_values = &#39;normalized&#39;,
                                cluster_column = &#39;leiden_clus&#39;)
# violinplot
topgenes_scran = scran_markers[, head(.SD, 2), by = &#39;cluster&#39;]$genes
violinPlot(mini_visium, genes = topgenes_scran, cluster_column = &#39;leiden_clus&#39;,
        strip_text = 10, strip_position = &#39;right&#39;)

# metadata heatmap
topgenes_scran = scran_markers[, head(.SD, 6), by = &#39;cluster&#39;]$genes
plotMetaDataHeatmap(mini_visium, selected_genes = topgenes_scran,
                metadata_cols = c(&#39;leiden_clus&#39;))
</pre></div>
</div>
</div>
<div class="section" id="cell-type-annotation">
<h2>6. Cell Type Annotation<a class="headerlink" href="#cell-type-annotation" title="Permalink to this headline">¶</a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">clusters_cell_types</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s1">&#39;Gfap_cells&#39;</span><span class="p">,</span> <span class="s1">&#39;Tbr1_cells&#39;</span><span class="p">,</span> <span class="s1">&#39;Tcf7l2_cells&#39;</span><span class="p">,</span> <span class="s1">&#39;Wfs1_cells&#39;</span><span class="p">,</span> <span class="s1">&#39;Nptxr_cells&#39;</span><span class="p">)</span>
<span class="n">names</span><span class="p">(</span><span class="n">clusters_cell_types</span><span class="p">)</span> <span class="o">=</span> <span class="mi">1</span><span class="p">:</span><span class="mi">5</span>
<span class="n">mini_visium</span> <span class="o">=</span> <span class="n">annotateGiotto</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">mini_visium</span><span class="p">,</span> <span class="n">annotation_vector</span> <span class="o">=</span> <span class="n">clusters_cell_types</span><span class="p">,</span>
                        <span class="n">cluster_column</span> <span class="o">=</span> <span class="s1">&#39;leiden_clus&#39;</span><span class="p">,</span> <span class="n">name</span> <span class="o">=</span> <span class="s1">&#39;cell_types&#39;</span><span class="p">)</span>

<span class="c1"># check new cell metadata</span>
<span class="n">pDataDT</span><span class="p">(</span><span class="n">mini_visium</span><span class="p">)</span>

<span class="c1"># visualize annotations</span>
<span class="n">spatDimPlot</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">mini_visium</span><span class="p">,</span> <span class="n">cell_color</span> <span class="o">=</span> <span class="s1">&#39;cell_types&#39;</span><span class="p">,</span> <span class="n">spat_point_size</span> <span class="o">=</span> <span class="mi">3</span><span class="p">,</span> <span class="n">dim_point_size</span> <span class="o">=</span> <span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="spatial-grid">
<h2>7. Spatial Grid<a class="headerlink" href="#spatial-grid" title="Permalink to this headline">¶</a></h2>
<div class="section" id="create-a-spatial-grid">
<h3>7.1 Create a Spatial Grid<a class="headerlink" href="#create-a-spatial-grid" title="Permalink to this headline">¶</a></h3>
<p>Create a grid based on defined stepsizes in the x,y(,z) axes.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">mini_visium</span> <span class="o">&lt;-</span> <span class="n">createSpatialGrid</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">mini_visium</span><span class="p">,</span>
                         <span class="n">sdimx_stepsize</span> <span class="o">=</span> <span class="mi">300</span><span class="p">,</span>
                         <span class="n">sdimy_stepsize</span> <span class="o">=</span> <span class="mi">300</span><span class="p">,</span>
                         <span class="n">minimum_padding</span> <span class="o">=</span> <span class="mi">50</span><span class="p">)</span>
<span class="n">showGrids</span><span class="p">(</span><span class="n">mini_visium</span><span class="p">)</span>
<span class="n">spatPlot</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">mini_visium</span><span class="p">,</span> <span class="n">show_grid</span> <span class="o">=</span> <span class="n">T</span><span class="p">,</span> <span class="n">point_size</span> <span class="o">=</span> <span class="mf">1.5</span><span class="p">)</span>

<span class="c1"># extract grid and associated metadata spots</span>
<span class="n">annotated_grid</span> <span class="o">=</span> <span class="n">annotateSpatialGrid</span><span class="p">(</span><span class="n">mini_visium</span><span class="p">)</span>
<span class="n">annotated_grid_metadata</span> <span class="o">=</span> <span class="n">annotateSpatialGrid</span><span class="p">(</span><span class="n">mini_visium</span><span class="p">,</span>
                                      <span class="n">cluster_columns</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s1">&#39;leiden_clus&#39;</span><span class="p">,</span> <span class="s1">&#39;cell_types&#39;</span><span class="p">,</span> <span class="s1">&#39;nr_genes&#39;</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="section" id="spatial-enrichment-cell-type-distribution">
<h3>7.2 Spatial Enrichment: Cell-Type Distribution<a class="headerlink" href="#spatial-enrichment-cell-type-distribution" title="Permalink to this headline">¶</a></h3>
<p>Here we will use known markers for different mouse brain cell types to identify which cell types are enriched in the individual spots or identified clusters.</p>
<p><a class="reference external" href="https://www.biorxiv.org/content/10.1101/294918v2">Paper: eisel, A. et al. Molecular Architecture of the Mouse Nervous System. Cell 174, 999-1014.e22 (2018).</a></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>## cell type signatures ##
## combination of all marker genes identified in Zeisel et al
sign_matrix_path = system.file(&quot;extdata&quot;, &quot;sig_matrix.txt&quot;, package = &#39;Giotto&#39;)
brain_sc_markers = data.table::fread(sign_matrix_path) # file don&#39;t exist in data folder
sig_matrix = as.matrix(brain_sc_markers[,-1]); rownames(sig_matrix) = brain_sc_markers$Event

## enrichment tests
mini_visium = runSpatialEnrich(mini_visium,
                        sign_matrix = sig_matrix,
                        enrich_method = &#39;PAGE&#39;) #default = &#39;PAGE&#39;

## heatmap of enrichment versus annotation (e.g. clustering result)
cell_types = colnames(sig_matrix)
plotMetaDataCellsHeatmap(gobject = mini_visium,
                        metadata_cols = &#39;leiden_clus&#39;,
                        value_cols = cell_types,
                        spat_enr_names = &#39;PAGE&#39;,
                        x_text_size = 8, y_text_size = 8)


enrichment_results = mini_visium@spatial_enrichment$PAGE
enrich_cell_types = colnames(enrichment_results)
enrich_cell_types = enrich_cell_types[enrich_cell_types != &#39;cell_ID&#39;]

## spatplot
spatCellPlot(gobject = mini_visium, spat_enr_names = &#39;PAGE&#39;,
        cell_annotation_values = enrich_cell_types,
        cow_n_col = 3,coord_fix_ratio = NULL, point_size = 1)
</pre></div>
</div>
</div>
</div>
<div class="section" id="spatial-network">
<h2>8. Spatial Network<a class="headerlink" href="#spatial-network" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Visualize information about the default Delaunay network</p></li>
<li><p>Create a spatial Delaunay network (default)</p></li>
<li><p>Create a spatial kNN network</p></li>
</ul>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">plotStatDelaunayNetwork</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">mini_visium</span><span class="p">,</span> <span class="n">maximum_distance</span> <span class="o">=</span> <span class="mi">300</span><span class="p">)</span>
<span class="n">mini_visium</span> <span class="o">=</span> <span class="n">createSpatialNetwork</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">mini_visium</span><span class="p">,</span> <span class="n">minimum_k</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">maximum_distance_delaunay</span> <span class="o">=</span> <span class="mi">400</span><span class="p">)</span>
<span class="n">mini_visium</span> <span class="o">=</span> <span class="n">createSpatialNetwork</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">mini_visium</span><span class="p">,</span> <span class="n">minimum_k</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s1">&#39;kNN&#39;</span><span class="p">,</span> <span class="n">k</span> <span class="o">=</span> <span class="mi">10</span><span class="p">)</span>
<span class="n">showNetworks</span><span class="p">(</span><span class="n">mini_visium</span><span class="p">)</span>

<span class="c1"># visualize the two different spatial networks</span>
<span class="n">spatPlot</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">mini_visium</span><span class="p">,</span> <span class="n">show_network</span> <span class="o">=</span> <span class="n">T</span><span class="p">,</span>
        <span class="n">network_color</span> <span class="o">=</span> <span class="s1">&#39;blue&#39;</span><span class="p">,</span> <span class="n">spatial_network_name</span> <span class="o">=</span> <span class="s1">&#39;Delaunay_network&#39;</span><span class="p">,</span>
        <span class="n">point_size</span> <span class="o">=</span> <span class="mf">2.5</span><span class="p">,</span> <span class="n">cell_color</span> <span class="o">=</span> <span class="s1">&#39;leiden_clus&#39;</span><span class="p">)</span>

<span class="n">spatPlot</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">mini_visium</span><span class="p">,</span> <span class="n">show_network</span> <span class="o">=</span> <span class="n">T</span><span class="p">,</span>
        <span class="n">network_color</span> <span class="o">=</span> <span class="s1">&#39;blue&#39;</span><span class="p">,</span> <span class="n">spatial_network_name</span> <span class="o">=</span> <span class="s1">&#39;kNN_network&#39;</span><span class="p">,</span>
        <span class="n">point_size</span> <span class="o">=</span> <span class="mf">2.5</span><span class="p">,</span> <span class="n">cell_color</span> <span class="o">=</span> <span class="s1">&#39;leiden_clus&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="spatial-genes">
<h2>9. Spatial Genes<a class="headerlink" href="#spatial-genes" title="Permalink to this headline">¶</a></h2>
<p>Identify spatial genes with 3 different methods:</p>
<ul class="simple">
<li><p>binSpect with kmeans binarization (default)</p></li>
<li><p>binSpect with rank binarization</p></li>
<li><p>silhouetteRank</p></li>
</ul>
<p>Visualize top 4 genes per method.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>km_spatialgenes = binSpect(mini_visium)
spatGenePlot(mini_visium, expression_values = &#39;scaled&#39;,
        genes = km_spatialgenes[1:4]$genes,
        point_shape = &#39;border&#39;, point_border_stroke = 0.1,
        show_network = F, network_color = &#39;lightgrey&#39;, point_size = 2.5,
        cow_n_col = 2)

rank_spatialgenes = binSpect(mini_visium, bin_method = &#39;rank&#39;)
spatGenePlot(mini_visium, expression_values = &#39;scaled&#39;,
        genes = rank_spatialgenes[1:4]$genes,
        point_shape = &#39;border&#39;, point_border_stroke = 0.1,
        show_network = F, network_color = &#39;lightgrey&#39;, point_size = 2.5,
        cow_n_col = 2)


silh_spatialgenes = silhouetteRank(gobject = mini_visium) # TODO: suppress print output
spatGenePlot(mini_visium, expression_values = &#39;scaled&#39;,
        genes = silh_spatialgenes[1:4]$genes,
        point_shape = &#39;border&#39;, point_border_stroke = 0.1,
        show_network = F, network_color = &#39;lightgrey&#39;, point_size = 2.5,
        cow_n_col = 2)
</pre></div>
</div>
</div>
<div class="section" id="spatial-co-expression-patterns">
<h2>10. Spatial Co-Expression Patterns<a class="headerlink" href="#spatial-co-expression-patterns" title="Permalink to this headline">¶</a></h2>
<p>Identify robust spatial co-expression patterns using the spatial network or grid and a subset of individual spatial genes.</p>
<div class="section" id="calculate-spatial-correlation-scores">
<h3>10.1 Calculate Spatial Correlation Scores<a class="headerlink" href="#calculate-spatial-correlation-scores" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span># 1. calculate spatial correlation scores
ext_spatial_genes = km_spatialgenes[1:100]$genes
spat_cor_netw_DT = detectSpatialCorGenes(mini_visium,
                                        method = &#39;network&#39;, spatial_network_name = &#39;Delaunay_network&#39;,
                                        subset_genes = ext_spatial_genes)
</pre></div>
</div>
</div>
<div class="section" id="cluster-correlation-scores">
<h3>10.2 Cluster Correlation Scores<a class="headerlink" href="#cluster-correlation-scores" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span># 2. cluster correlation scores
spat_cor_netw_DT = clusterSpatialCorGenes(spat_cor_netw_DT, name = &#39;spat_netw_clus&#39;, k = 8)
heatmSpatialCorGenes(mini_visium, spatCorObject = spat_cor_netw_DT, use_clus_name = &#39;spat_netw_clus&#39;)


netw_ranks = rankSpatialCorGroups(mini_visium, spatCorObject = spat_cor_netw_DT, use_clus_name = &#39;spat_netw_clus&#39;)
top_netw_spat_cluster = showSpatialCorGenes(spat_cor_netw_DT, use_clus_name = &#39;spat_netw_clus&#39;,
                                    selected_clusters = 6, show_top_genes = 1)

cluster_genes_DT = showSpatialCorGenes(spat_cor_netw_DT, use_clus_name = &#39;spat_netw_clus&#39;, show_top_genes = 1)
cluster_genes = cluster_genes_DT$clus; names(cluster_genes) = cluster_genes_DT$gene_ID

mini_visium = createMetagenes(mini_visium, gene_clusters = cluster_genes, name = &#39;cluster_metagene&#39;)
        spatCellPlot(mini_visium,
        spat_enr_names = &#39;cluster_metagene&#39;,
        cell_annotation_values = netw_ranks$clusters,
        point_size = 1.5, cow_n_col = 3)
</pre></div>
</div>
</div>
</div>
<div class="section" id="spatial-hmrf-domains">
<h2>11. Spatial HMRF Domains<a class="headerlink" href="#spatial-hmrf-domains" title="Permalink to this headline">¶</a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>hmrf_folder = paste0(temp_dir,&#39;/&#39;,&#39;11_HMRF/&#39;)
if(!file.exists(hmrf_folder)) dir.create(hmrf_folder, recursive = T)

# perform hmrf
my_spatial_genes = km_spatialgenes[1:100]$genes
HMRF_spatial_genes = doHMRF(gobject = mini_visium,
                        expression_values = &#39;scaled&#39;,
                        spatial_genes = my_spatial_genes,
                        spatial_network_name = &#39;Delaunay_network&#39;,
                        k = 8,
                        betas = c(28,2,2),
                        output_folder = paste0(hmrf_folder, &#39;/&#39;, &#39;Spatial_genes_brain/SG_top100_k8_scaled&#39;))

# check and select hmrf
for(i in seq(28, 30, by = 2)) {
viewHMRFresults2D(gobject = mini_visium,
                HMRFoutput = HMRF_spatial_genes,
                k = 8, betas_to_view = i,
                point_size = 2)
}

mini_visium = addHMRF(gobject = mini_visium,
                HMRFoutput = HMRF_spatial_genes,
                k = 8, betas_to_add = c(28),
                hmrf_name = &#39;HMRF&#39;)

giotto_colors = getDistinctColors(8)
names(giotto_colors) = 1:8
spatPlot(gobject = mini_visium, cell_color = &#39;HMRF_k8_b.28&#39;,
        point_size = 3, coord_fix_ratio = 1, cell_color_code = giotto_colors)
</pre></div>
</div>
<div class="section" id="export-giotto-analyzer-to-viewer">
<h3>12. Export Giotto Analyzer to Viewer<a class="headerlink" href="#export-giotto-analyzer-to-viewer" title="Permalink to this headline">¶</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">viewer_folder</span> <span class="o">=</span> <span class="n">paste0</span><span class="p">(</span><span class="n">temp_dir</span><span class="p">,</span> <span class="s1">&#39;/&#39;</span><span class="p">,</span> <span class="s1">&#39;Mouse_cortex_viewer&#39;</span><span class="p">)</span>

<span class="c1"># select annotations, reductions and expression values to view in Giotto Viewer</span>
<span class="n">exportGiottoViewer</span><span class="p">(</span><span class="n">gobject</span> <span class="o">=</span> <span class="n">mini_visium</span><span class="p">,</span> <span class="n">output_directory</span> <span class="o">=</span> <span class="n">viewer_folder</span><span class="p">,</span>
                <span class="n">factor_annotations</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s1">&#39;cell_types&#39;</span><span class="p">,</span>
                                  <span class="s1">&#39;leiden_clus&#39;</span><span class="p">,</span>
                                  <span class="s1">&#39;HMRF_k8_b.28&#39;</span><span class="p">),</span>
                <span class="n">numeric_annotations</span> <span class="o">=</span> <span class="s1">&#39;total_expr&#39;</span><span class="p">,</span>
                <span class="n">dim_reductions</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s1">&#39;umap&#39;</span><span class="p">),</span>
                <span class="n">dim_reduction_names</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="s1">&#39;umap&#39;</span><span class="p">),</span>
                <span class="n">expression_values</span> <span class="o">=</span> <span class="s1">&#39;scaled&#39;</span><span class="p">,</span>
                <span class="n">expression_rounding</span> <span class="o">=</span> <span class="mi">3</span><span class="p">,</span>
                <span class="n">overwrite_dir</span> <span class="o">=</span> <span class="n">T</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="mini_3D_STARmap.html" class="btn btn-neutral float-right" title="mini3D STARmap" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="mini_seqFISH.html" class="btn btn-neutral float-left" title="mini seqFISH" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2021, Ruben Dries, Qian Zhu, Huipeng Li, Rui Dong, Guo-Cheng Yuan.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>